File: biopython_rewrite/blast_parser.py
#!/usr/bin/env python3
# Parse BLAST tabular output and report best hit per query
import sys
from collections import defaultdict
if len(sys.argv)<2:
    print('Usage: blast_parser.py blast.tsv'); sys.exit(1)
best = {}
with open(sys.argv[1]) as fh:
    for line in fh:
        parts = line.strip().split('\t')
        q,s,pid,length,mm,gap,qstart,qend,sstart,send,evalue,bits = parts[:12]
        bits = float(bits)
        if q not in best or bits > best[q]['bits']:
            best[q] = {'subject':s,'pid':pid,'bits':bits,'evalue':evalue}
for q in sorted(best):
    b = best[q]
    print('\t'.join([q,b['subject'],b['pid'],str(b['bits']),b['evalue']]))

File: legacy_perl/blast_parser.pl
#!/usr/bin/env perl
# Parse BLAST tabular output and report best hit per query
use strict;
use warnings;
my $file = shift or die "Usage: $0 blast.tsv\n";
open my $fh, '<', $file or die $!;
my %best;
while(<$fh>){
    chomp;
    my ($q,$s,$pid,$len,$mismatch,$gap,$qstart,$qend,$sstart,$send,$evalue,$bits) = split /\t/;
    if(!exists $best{$q} || $bits > $best{$q}{bits}){
        $best{$q} = {subject=>$s, pid=>$pid, bits=>$bits, evalue=>$evalue};
    }
}
foreach my $q (sort keys %best){
    print join("\t", $q, $best{$q}{subject}, $best{$q}{pid}, $best{$q}{bits}, $best{$q}{evalue}), "\n";
}
